### PR TITLE
pkg: check file stats

### PR DESCRIPTION
**Pull request by spzala**
modify file util.

### CODE REVIEW

File: pkg/fileutil/fileutil.go [hunk 1]
```diff
@@ -104,3 +112,22 @@ func ZeroToEnd(f *os.File) error {
 	_, err = f.Seek(off, io.SeekStart)
 	return err
 }
+
+// CheckPermission checks permission on an existing dir or a file.
+// Returns error if dir or file empty or exist with a different permission than specified.
+func CheckPermission(dir string, perm os.FileMode) error {
```
Comment (2020758) [inline pkg/fileutil/fileutil.go:118]: checkDirPremission?

Comment (xiang90) [inline pkg/fileutil/fileutil.go:118]: CheckDirPermission?

**Comment by xiang90** [comment]
LGTM after improving the func naming

File: pkg/fileutil/fileutil.go (hunk 1 above)
Comment (spzala) [reply pkg/fileutil/fileutil.go:118]: @xiang90 sure, sounds good, thanks!

**Comment by spzala** [comment]
> LGTM after improving the func naming

Done :-) Thanks @xiang90

**Comment by xiang90** [comment]
@spzala Thank you!

**Comment by cretz** [comment]
This change is causing issues for me for embedded etcd on Windows. Specifically at https://github.com/etcd-io/etcd/blob/732df43cf85b89ffb5d40d0226206672e4a8be25/embed/etcd.go#L416. Shall I open a new issue?

**Comment by spzala** [comment]
> This change is causing issues for me for embedded etcd on Windows. Specifically at
> 
> https://github.com/etcd-io/etcd/blob/732df43cf85b89ffb5d40d0226206672e4a8be25/embed/etcd.go#L416
> 
> Getting `directory "fixtures\\peer" exist without desired file permission. "drwxrwxrwx"`. Shall I open a new issue?
> 
> EDIT: This same check is done w/ the data dir too. At the least nix-style dir permission checks should only occur on nix platforms. This also is probably a good time to consider Windows Travis environments.

@cretz thanks for reporting, and yes please open a new issue. Also, if you can provide a fix with a PR that would be great and much appreciated. I have no way to test on windows, but if you (someone from your side) can't work on creating a PR please let me know I will work on it? Thanks!  (p.s. the Windows Travis would be a separate work and we would need some help there too)

**Comment by cretz** [comment]
Thanks. Created https://github.com/etcd-io/etcd/issues/11920 ...might not have the cycles for a PR for a bit...

**Comment by spzala** [comment]
> Thanks. Created #11920 ...might not have the cycles for a PR for a bit...

Thanks @cretz

**Comment by polyrabbit** [comment]
@spzala @cretz Please see my fix #11945

**Comment by spzala** [comment]
> @spzala @cretz Please see my fix #11945

@polyrabbit thanks, I have put the review comments in the PR. Hopefully we will have it merged soon.

File: etcdmain/etcd.go [hunk 2]
```diff
@@ -260,9 +260,16 @@ func startProxy(cfg *config) error {
 	}
 
 	cfg.ec.Dir = filepath.Join(cfg.ec.Dir, "proxy")
-	err = os.MkdirAll(cfg.ec.Dir, fileutil.PrivateDirMode)
-	if err != nil {
-		return err
+	if fileutil.Exist(cfg.ec.Dir) {
+		err := fileutil.CheckDirPermission(cfg.ec.Dir, fileutil.PrivateDirMode)
+		if err != nil {
+			return err
```
Comment (nordicdyno) [inline etcdmain/etcd.go:266]: > fmt.Errorf("%w: expect permissions %v", err, fileutil.PrivateDirMode)

will be much friendlier for end users

Comment (spzala) [reply etcdmain/etcd.go:266]: It's being addressed in here https://github.com/etcd-io/etcd/pull/11997 Please feel free to add review comments there. Thanks!

**Pull request merged**